src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            Contract.Assert(_item != null);
+
+            _items[index] = _item;
+        }
+
+        /// <summary>
+        /// Copy changed values is an implementation of Patch
+        /// </summary>
+        /// <param name="original">Original collection of the Type which needs to be updated</param>              
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        internal EdmChangedObjectCollection<TStructuralType> CopyChangedValues(ICollection<TStructuralType> original)
+        {
+            //Here we need to Find the key of the Type, then only we will be able to find from the collection that which item in the collection
+            //corresponds to the item in delta list(Edmchangedobjectcoll).
+            Type type = typeof(TStructuralType);
+            IEnumerable<IEdmStructuralProperty> keys = EntityType.Key();
+
+            EdmChangedObjectCollection<TStructuralType> edmChangedObjectCollection = new EdmChangedObjectCollection<TStructuralType>(EntityType);
+
+            foreach (dynamic changedObj in _items)
+            {
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+                IEdmChangedObject edmChangedObject = null;
+               
+                //Get filtered item based on keys
+                TStructuralType originalObj = GetFilteredItem(type, keys, original, changedObj); ;
+
+                try
+                {
+                    IEdmDeltaDeletedEntityObject deletedObj = changedObj as IEdmDeltaDeletedEntityObject;
+
+                    if (deletedObj != null)
+                    {
+                        if (deletedObj.Reason == DeltaDeletedEntryReason.Deleted)
+                        {
+                            operation = DataModificationOperationKind.Delete;
+                        }
+                        else
+                        {
+                            operation = DataModificationOperationKind.Unlink;
+                        }
+
+                        if (originalObj != null)
+                        {
+                            //This case handle deletions
+                            original.Remove(originalObj);
+                        }
+                    }
+                    else
+                    {
+                        if (originalObj == null)
+                        {
+                            operation = DataModificationOperationKind.Insert;
+                            //This case handle additions
+                            originalObj = Activator.CreateInstance(changedObj.ExpectedClrType);
+                            original.Add(originalObj);
+                        }
+
+                        //Patch for addition/update. This will call Delta<T> for each item in the collection
```

Comment:
>addition [](start = 36, length = 8)

Note that this assumes that the original collection keeps a reference to the added object as opposed to copying it (and accepts adding an empty instance).  If we don't think that will always be the case, we should Patch the new object created in 177 before adding it to the original list in 178 (and just handle the update case here by wrapping 181-182 in an else{}.